### PR TITLE
Align scaffolding with K8up

### PR DESCRIPTION
**Pull request by ccremer**
## Summary

* upgrades Espejo to K8s 1.20, with controller-runtime v0.8.3 and Go 1.16
* Replaces own SyncConfigCondition with metav1.Condition
* CI/CD:
  * Copy workflows from K8up, including changelog generator
  * CodeClimate initialized
  * Copy `Makefile` from K8up
* Testing:
  * Migrate previous e2e test cases to bats-detik
  * Migrate integration tests with Ginkgo/Gomega to plain `testing`
* Add `apiextensions.k8s.io/v1` CRD version.
* Change default docker user from 1000:0 to 65532:65532
* Kustomize
  * Remove rbac-auth proxy
  * Add RBAC rules for leader election (K8s >= 1.16)
* Update readme

## Checklist

- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`,
      as they show up in the changelog
- [x] Update the documentation.
- [x] Update tests.

### CODE REVIEW

File: api/v1alpha1/syncconfig_types.go [hunk 1]
```diff
@@ -22,6 +14,7 @@ type (
 		// NamespaceSelector defines which namespaces should be targeted
 		NamespaceSelector *NamespaceSelector `json:"namespaceSelector,omitempty"`
 		// SyncItems lists items to be synced to targeted namespaces
+		// +kubebuilder:pruning:PreserveUnknownFields
```
Comment (srueg) [inline api/v1alpha1/syncconfig_types.go:17]:
```suggestion
		// +kubebuilder:pruning:PreserveUnknownFields
		// +kubebuilder:validation:EmbeddedResource
```

Would this maybe help to get rid of the mentioned validation errors?
https://book.kubebuilder.io/reference/markers/crd-validation.html

File: api/v1alpha1/syncconfig_types.go [hunk 2]
```diff
@@ -66,14 +59,14 @@ type (
 
 	// SyncConfigCondition describes a status condition of a SyncConfig
 	SyncConfigCondition struct {
```
Comment (srueg) [inline api/v1alpha1/syncconfig_types.go:61]: I guess we can remove this struct since using `metav1.Condition`?

**Review by srueg — CHANGES_REQUESTED** [review]

File: api/v1alpha1/syncconfig_types.go (hunk 1 above)
Comment (ccremer) [reply api/v1alpha1/syncconfig_types.go:17]: uuuh, lemme try

Comment (ccremer) [reply api/v1alpha1/syncconfig_types.go:17]: I think it's having a problem with this property being an array:
```
The CustomResourceDefinition "syncconfigs.sync.appuio.ch" is invalid: spec.validation.openAPIV3Schema.properties[spec].properties[syncItems].type: Invalid value: "array": must be object if x-kubernetes-embedded-resource is true
```
```
		// +kubebuilder:pruning:PreserveUnknownFields
		// +kubebuilder:validation:EmbeddedResource

		// SyncItems lists items to be synced to targeted namespaces
		SyncItems []unstructured.Unstructured `json:"syncItems,omitempty"`
```

Comment (srueg) [reply api/v1alpha1/syncconfig_types.go:17]: We might be able to define a type alias for `unstructured.Unstructured` and set the validation on it?

Comment (ccremer) [reply api/v1alpha1/syncconfig_types.go]: I think I was able to get rid of the validation error by aliasing `Unstructured` and put the preserveUnknownFields marker there.

Comment (srueg) [reply api/v1alpha1/syncconfig_types.go]: Can you try with `// +kubebuilder:validation:EmbeddedResource` as well? Would be nice to have some basic validation that the sync items include valid K8s objects.

Comment (ccremer) [reply api/v1alpha1/syncconfig_types.go]: I put that marker in both `SyncItems []SyncItem` and `SyncItem`. `make generate` did not add any other fields to the CRD spec except for preserveUnknownFields, so I left it out again.
Do you think we should do it for forward compatibility?

Comment (ccremer) [reply api/v1alpha1/syncconfig_types.go]: I left it in there now, even it doesn't do anything now.
`// +kubebuilder:validation:EmbeddedResource` may be working if `Unstructured` is used as its own field, but not in an array.

**Review by srueg — APPROVED** [review]

**Pull request merged**